R/ic.R

```diff
@@ -5,66 +5,54 @@
 #' @return If `x` is an expression, returns the result of evaluating `x`. If `x` is missing nothing
 #'   is returned.
 #'
-#' @export
-#'
 #' @examples
 #' f <- function(x) x < 0
 #'
 #' ic(f(1))
 #'
 #' ic(f(-1))
+#'
+#' @importFrom rlang enquo quo_is_missing trace_back quo_get_expr caller_fn caller_env expr_deparse eval_tidy fn_env env_label maybe_missing
+#' @importFrom glue glue
+#' @export
 ic <- function(x) {
   # Capture the input to allow us to work with the expression and value separately
-  q <- rlang::enquo(x)
+  q <- enquo(x)
 
   # The behaviour of the function changes depending on whether input is provided or not.
-  missing_input <- rlang::quo_is_missing(q)
+  missing_input <- quo_is_missing(q)
 
   # In the event that icecream is totally disabled we will just return the input.
-  #
-  # TODO: This triggers evaluation of the input. Is this a problem?
-  if (isFALSE(getOption("icecream.enabled"))) {
-    if (!missing_input) {
-      return(x)
+  if (getOption("icecream.enabled")) {
+    trace <- trace_back()
+    num_calls <- length(trace$calls)
+
+    parent_ref <-  if (num_calls > 1) trace$calls[[num_calls - 1]][[1]] else NULL
+    ref <- attr(trace$calls[[num_calls]], "srcref")
+    loc <- src_loc(ref)
+
+    # Case when location of file is unavailable
+    if (nchar(loc) == 0) {
+      # Probs want to look at environments
+      caller <- caller_fn()
+      caller_env <- if (is.null(caller)) caller_env() else fn_env(caller)
+
+      loc <- env_label(caller_env)
+      loc <- glue("<env: {loc}>")
     }
-    return(invisible())
-  }
 
-  trace <- rlang::trace_back()
-  num_calls <- length(trace$calls)
-  parent_ref <-  if (num_calls > 1) trace$calls[[num_calls - 1]][[1]] else NULL
-  ref <- attr(trace$calls[[num_calls]], "srcref")
-  loc <- src_loc(ref)
-  if (nchar(loc) == 0) {
-    # Probs want to look at environments
-    caller_fn <- sys.function(-1)
-    if (is.null(caller_fn)) {
-      e <- sys.frame(-1)
+    # If we have inputs then we want the expression and value to be included in the context
+    # object as well.
+    if (!missing_input) {
+      deparsed_expression <- expr_deparse(quo_get_expr(q))
+      x <- eval_tidy(q)
+      ic_print(loc, parent_ref, deparsed_expression, x)
     } else {
-      e <- environment(caller_fn)
+      ic_print(loc, parent_ref)
     }
-    if (is.environment(e)) {
-      loc <- rlang::env_label(e)
-      loc <- glue::glue("<env: {loc}>")
-    }
-  }
-
-  # If we have inputs then we will want the expression and value to be included in the context
-  # object as well.
-  expression <- NULL
-  value <- NULL
-  if (!missing_input) {
-    expression <- deparse(rlang::quo_get_expr(q))
-    value <- rlang::eval_tidy(q)
   }
-
-  # Print the output!
-  ic_print(loc, parent_ref, expression, value)
-
   # Return the result
-  if (!missing_input) {
-    return(value)
-  }
+  if (missing_input) invisible() else invisible(x)
```

Comment:
I don't like visible return for one simple reason: we use `ic` to see the value of the wrapped expression. If we don't wrap the value, we see it anyway. If, on the other hand, we use `ic` and want to return a value (as in returning from a function, for example), when we call the code we will have the value of the expression printed out twice: once from the message generated by `ic`, and the second time as simply printing the returned value to the screen.

I understand `ic` as a replacement for print, as a function that can be used instead of `print` -- and `print` also returns the value of the expression being printed out invisibly.

I'm trying to think of some use cases where this invisible return would really be undesirable, and only two come to mind:

* `some_fun(ic(some_expr))` where `some_func` somehow checks the visibility of the argument being taken
* `ic` wraps the expression returned by the function, `ic` is enabled and as a result the function returns

The first case seems extremely rare (if possible at all) to me.  The second case, on the other hand, is more significant, and you need to think about what to do about it.

Reply:
Agreed, the comparison to `print()` makes sense so let's stick with `invisible()`. 

I'm not sure what you mean by your second point, could you give an example? I can't think of any circumstances where `ic()` would cause a function to return unexpectedly.

``` r
f <- function(x) {
  y <- x + 1
  ic(y)  # This will return, but so would `y` on its own
}

g <- function(x) {
  y <- x + 1
  ic(y) # This won't return, just as `y` on its own wouldn't
  y + 1
}
```

Am I missing something?

Reply:
Okay, sorry, my bad -- I had edited my response before I published it and changed something in this case. The sentence isn't even complete. I really cannot recreate what I had in my mind then.

But I also realized that when disabled `ic` should return visibly, cause now it always returns invisibly.

Reply:
Nah that's all good, as long as I'm not missing anything! Good point about the visibility when `ic()` is disabled. I think we can handle that by reworking `ic()` along these lines:

``` r
ic <- function(x) {
  q <- enquo(x)
  missing_input <- quo_is_missing(q)
  if (!getOption("icecream.enabled") && !missing_input) {
    return(eval_tidy(q))
  }
  # ... everything else...
} 
```
I have a feeling there's  a neater way of handling that but it seems to work. How busy are you? If you have the time and inclination to make that change and update the PR I think we will be good to merge. Otherwise I'm happy to merge this PR as-is and I'll sort it myself. I'm away for the next few days so no rush.

Reply:
We can also just add `else` branch at the end of `if (getOption("icecream.enabled"))` and additional checking of `if (!missing_input)`. I incorporated this in my next commit. The result is the same.